design.md

```diff
@@ -0,0 +1,109 @@
+* The client/cli is an interface for users to start, stop and get the status of a running job on a remote server.
+
+## Worker library API
+The worker library consists of 2 components. The supervisor and the process.
+### Supervisor API
+The supervisor is responsible for managing all jobs in the system. It provides apis to start, stop, get the output and get the status of a job. 
+* `Newsupervisor()` - Returns a new instance of a supervisor
+* `Start(cmd string, args []string, opts...Options)` - Starts a process and return a unique job id. The job id is generated by the supervisor.
+* `Stop(id string)` - Stops a running process by sending the process group a `SIGTERM` signal.
+* `Run()` - Run the supervisor and listen for incoming job requests.
+* `WatchOutput(id string, c chan ProcessOutput)` - Add an output channel to subscribe to process output. All previous output, if any, from the process is replayed before sending new messages to the channel.
+* `UnWatchOutput(id string, c chan ProcessOutput)` - Unsubscribe to the process output.
+* `GetOutput(id string)` - Get all lines of output from the job's line buffer. Useful when getting the output of a finished job.
+* `Status(id string)` - Returns the status of a process
+
+
+## Process API
+`Process` is a wrapper around `os/exec.Cmd`
+* `NewProcess(name string, args []string, opts ...Options)` - Returns a new instance of a process given a command name and it's arguments. An optional list of options can also be specified.
+* `Start()` - Starts the process and returns a channel the caller can listen on to get the final status.
+* `Stop()` - Stops the process by sending a `SIGTERM` signal to it's process group.
+* `Signal(sig syscall.Signal)` - Send an os signal to the process group.
+* `Status()` - Get the current status of a process.
+* `IsInitialState()` - true if the process is in the initial state (the process isn't started yet).
+* `IsRunningState()` - true if the process is starting or running.
+* `IsFinalState()` - true if the process is in a final state
+
+### Output streaming
+Each process has a `StreamOutput` for managing process output. `StreamOutput` is an `io.Writer` that has a channel where new lines of output from the process are sent to. The caller should use this channel to listen for output from the process. The supervisor also stores each line of output comming from the `StreamOutput` for each job. These can be retrieved at any time by calling `GetOutput(...)`
+
+### Process options
+`Options` is a functional option that can be provided to `NewProcess(...)` to customize the process. The only options provided are `WithStdOutWriters(writers...io.Writer)` and `WithStdErrWriters(writers...io.Writer)`. These provide additional writers to the process standard output and standard error respectively. By default all process output goes to the `StreamBuffer`
```

Comment:
If the gRPC API will not use these options, let's remove them

Reply:
Ok will do